openssl/thread_setup.c

```diff
@@ -0,0 +1,46 @@
+//go:build linux && !android
+// +build linux,!android
+
+#include "goopenssl.h"
+
+#include <stdio.h>
+#include <pthread.h>
+#include <sys/types.h>
+#include <sys/syscall.h>
+
+#define _GNU_SOURCE
+#include <unistd.h>
+ 
+#define CRYPTO_LOCK      0x01
+
+/* This array will store all of the mutexes available to OpenSSL. */ 
+static pthread_mutex_t *mutex_buf = NULL;
+ 
+static void locking_function(int mode, int n, const char *file, int line)
+{
+    if(mode & CRYPTO_LOCK) {
+        pthread_mutex_lock(&mutex_buf[n]);
+    } else {
+        pthread_mutex_unlock(&mutex_buf[n]);
+    }
+}
+ 
+static unsigned long id_function(void)
+{
+    return (unsigned long)syscall(__NR_gettid);
+}
+ 
+int go_openssl_thread_setup(void)
+{
+    mutex_buf = malloc(go_openssl_CRYPTO_num_locks()*sizeof(pthread_mutex_t));
+    if(!mutex_buf) {
+        return 0;
+    }
+    int i;
+    for(i = 0;  i < go_openssl_CRYPTO_num_locks(); i++) {
+        pthread_mutex_init(&mutex_buf[i], NULL);
+    }
```

Comment:
A few changes to match the style in `openssl/goopenssl.c`:

```suggestion
    for (i = 0; i < go_openssl_CRYPTO_num_locks(); i++)
        pthread_mutex_init(&mutex_buf[i], NULL);
```

(And similar for the other blocks in the file.)

I personally prefer the style of always including `{` and `}` even for one statement, but one way or another, we should probably pick one and stick to it?